# Examples and Exercises from Think Stats, 2nd Edition

http://thinkstats2.com

Copyright 2016 Allen B. Downey

MIT License: https://opensource.org/licenses/MIT


In [1]:
from __future__ import print_function, division
import nsfg

In [14]:
import pandas as pd
from string import Template
import pyodbc

sql_conn = pyodbc.connect('''DRIVER={ODBC Driver 13 for SQL Server};
                            SERVER=zye0dz8stg.database.windows.net;
                            DATABASE=TimeDB;
                            UID=netwovensa;
                            PWD=Reallongpass1''') 
query = '''select FullName, ProjectName, Date=CAST(Date AS DATE),   
Billable=SUM(ActualWorkBillable), c.ClassName
	from NWTA_TimeSheet t
		join NWTA_Resource r
			on r.ResourceUID = t.OwnerResourceUID
		join NWTA_TimesheetLine l
			on l.TimesheetUID = t.TimesheetUID

		join NWTA_TimesheetClass c
			on l.ClassID = c.TimesheetClassID

		join NWTA_TimesheetActual a
			on a.TimesheetLineUID = l.TimesheetLineUID
		join NWTA_Assignment n
			on n.AssignmentUID = l.AssignmentUID
		join NWTA_Task s
			on s.TaskUID = n.TaskUID
		join NWTA_Project pr
			on pr.ProjectUID = s.ProjectUID	

		where ActualWorkBillable > 0
		GROUP by r.FullName, ProjectName, Date, ClassName
	order by r.FullName, Date'''
df = pd.read_sql(query, sql_conn)
df['year'] =  pd.to_datetime(df['Date']).dt.year
p = df.pivot_table( index='FullName', columns='year', values='Billable', aggfunc=sum, fill_value=0)
p.to_excel('c:/hours.xlsx')
df.head(3)

# mydf1 = pd.read_csv('mydata/test.txt')
# sql = '''select item, brand, model, qty, price, "date"
#     from bill_memo key join bill_memo_product
#         key join inv_main
#             key join product
#                 order by "date" '''
# def getconn(db):
#     return sqlanydb.connect(uid='dba', pwd='sql', eng='server', dbn=db, host='kushserver' )

# def get_amount(row):
#     a=  (float(row['qty']) * float(row['price']))
#     return(a)
#     #return ("{:,}".format(a)) 

# pd.options.display.float_format = '{:,.0f}'.format

# dfa = pd.read_sql_query(sql, getconn('capi2013'))
# dfa['month'] = pd.to_datetime(dfa['date']).dt.month
# dfa['year'] =  pd.to_datetime(dfa['date']).dt.year
# dfa['amount'] =  dfa.apply(get_amount, axis=1) # return ("{:,}".format(number)) 
# dfa

FullName   ProjectName        Date  Billable ClassName  year
0  Abdul Sadat Mohammad  Marvell 2015  2015-02-09       8.0  Billable  2015
1  Abdul Sadat Mohammad  Marvell 2015  2015-02-10       8.0  Billable  2015
2  Abdul Sadat Mohammad  Marvell 2015  2015-02-11       8.0  Billable  2015

In [54]:
import pandas as pd
import sqlanydb
from string import Template

mydf1 = pd.read_csv('mydata/test.txt')
sql = '''select item, brand, model, qty, price, "date"
    from bill_memo key join bill_memo_product
        key join inv_main
            key join product
                order by "date" '''
def getconn(db):
    return sqlanydb.connect(uid='dba', pwd='sql', eng='server', dbn=db, host='kushserver' )

def get_amount(row):
    return float(row['qty']) * float(row['price'])

pd.options.display.float_format = '{:,.0f}'.format

dfa = pd.read_sql_query(sql, getconn('capi2013'))
dfa['month'] = pd.to_datetime(dfa['date']).dt.month
dfa['year'] =  pd.to_datetime(dfa['date']).dt.year
dfa['amount'] =  dfa.apply(get_amount, axis=1)

dfb = pd.read_sql_query(sql, getconn('capi2014'))
dfb['month'] = pd.to_datetime(dfb['date']).dt.month
dfb['year'] =  pd.to_datetime(dfb['date']).dt.year
dfb['amount'] =  dfb.apply(get_amount, axis=1)

dfc = pd.read_sql_query(sql, getconn('capi2015'))
dfc['month'] = pd.to_datetime(dfc['date']).dt.month
dfc['year'] =  pd.to_datetime(dfc['date']).dt.year
dfc['amount'] =  dfc.apply(get_amount, axis=1)

df1 = pd.read_sql_query(sql, getconn('capi2016'))
df1['month'] = pd.to_datetime(df1['date']).dt.month
df1['year'] =  pd.to_datetime(df1['date']).dt.year
df1['amount'] =  df1.apply(get_amount, axis=1)

df2 = pd.read_sql_query(sql, getconn('capi2017'))
df2['month'] = pd.to_datetime(df2['date']).dt.strftime('%Y-%m')
df2['year'] =  pd.to_datetime(df2['date']).dt.year
df2['month'] = pd.to_datetime(df2['date']).dt.month
df2['amount'] =  df2.apply(get_amount, axis=1)

df3 = pd.read_sql_query(sql, getconn('capi2018'))
df3['month'] = pd.to_datetime(df3['date']).dt.strftime('%Y-%m')
df3['year'] =  pd.to_datetime(df3['date']).dt.year
df3['month'] = pd.to_datetime(df3['date']).dt.month
df3['amount'] =  df3.apply(get_amount, axis=1)

df4 = pd.read_sql_query(sql, getconn('capi2019'))
df4['month'] = pd.to_datetime(df4['date']).dt.strftime('%Y-%m')
df4['year'] =  pd.to_datetime(df4['date']).dt.year
df4['month'] = pd.to_datetime(df4['date']).dt.month
df4['amount'] =  df4.apply(get_amount, axis=1)

df = [dfa,dfb,dfc,df1,df2,df3,df4]
df = pd.concat(df)

p = df.pivot_table( index='item', columns='year', values='amount', aggfunc=sum, fill_value=0)
p.to_excel('c:/sales_amount.xlsx')
#p.plot(kind='line', x='item', y='year')
p.head(10000)


year            2013      2014      2015      2016      2017      2018    2019
item                                                                          
 FREEZER           0         0         0         0    18,253         0       0
 PURIFIER          0         0         0         0         0    15,848       0
*****              0         0         0     4,367   100,000   187,725 150,000
AC         1,426,453 2,591,607 1,560,146 1,744,716 1,579,128   712,805 636,179
ADAPTER        4,453         0         0     2,032     5,983         0     999
ADAPTOR        7,663     1,143     9,679     5,422         0         0   1,144
AIRFRYER           0         0         0    10,480         0         0       0
ALEXA              0         0         0         0         0    24,530       0
ANTIVIRUS     23,666    49,450    47,532    25,429    20,901    11,229       0
AUDIO        231,030   230,670   174,760    79,258    38,331     5,424  38,433
AVCABLE            0         0       306     1,135         0         0       0
B. COOLER          0    25,153         0         0         0         0       0
B.PACK             0         0    10,044         0         0         0       0
B/GRIP             0         0         0    16,769         0         0       0
BACKPACK           0         0         0     4,629     3,774       720       0
BAG           24,122     4,838     2,095         0       339         0  22,331
BALLHEAD           0         0         0         0         0         0   7,966
BAT/CHAG     120,348   183,845   140,936    55,361    67,771    55,210  33,263
BAT/GRIP           0         0         0         0    15,625    39,153       0
BATTARY            0         0         0         0         0         0   3,390
BATTERY      540,774   531,402   644,239   425,059   294,867   283,824 198,452
BDHTSYSTEM    44,978         0         0         0         0         0       0
BIGTV              0         0         0         0         0         0       0
BINOCULAR    369,738   229,607   249,694   296,017   323,377   287,026 166,871
BLENDER            0     2,052         0     1,397     2,726     3,737     932
BLUETOOTH          0         0     8,333     5,932         0    43,203  30,170
BODYCAP            0         0       262         0         0         0       0
BOTTLE INK         0         0         0         0         0     1,758       0
BP/MONITOR         0         0         0         0         0     3,304   1,653
BPMONITOR          0         0         0         0         0     3,768   1,780
...              ...       ...       ...       ...       ...       ...     ...
TORCH            649         0         0         0         0         0       0
TRIMER        16,332     5,790     4,087         0         0         0       0
TRIMMER            0         0         0     4,716    42,681   139,620  76,110
TRIPOD        65,231    54,306    75,746    32,157    61,528    90,787  36,737
TRIPOTHEAD         0         0         0     7,424     8,297         0       0
TROLLEY            0         0       786         0         0         0       0
TVTUNER            0     2,429     6,885     3,571         0         0       0
UPS            6,333     9,190    38,000     6,952     2,190     1,864       0
USB                0         0     1,305     2,095     1,021     1,639   1,954
USBCABLE           0         0     2,283       333       547    13,425  11,616
USBCHARGER     4,107    27,131   146,095    17,749    46,736    27,290   4,788
USBEX              0         0         0         0       540       212     170
USBHUB         1,571       419     1,809       714       476       517       0
UV                 0         0         0         0         0         0   7,424
UVFILTER     188,462   196,497   184,696   204,701   315,765   372,487 169,587
V/C                0    21,834    22,271         0         0         0       0
V/CAM              0         0         0         0   178,166    50,625       0
V/CAMARA           0         0         0         0

## Examples from Chapter 1

Read NSFG data into a Pandas DataFrame.

In [2]:
preg = nsfg.ReadFemPreg()
preg.head()

caseid  pregordr  howpreg_n  howpreg_p  moscurrp  nowprgdk  pregend1  \
0       1         1        NaN        NaN       NaN       NaN       6.0   
1       1         2        NaN        NaN       NaN       NaN       6.0   
2       2         1        NaN        NaN       NaN       NaN       5.0   
3       2         2        NaN        NaN       NaN       NaN       6.0   
4       2         3        NaN        NaN       NaN       NaN       6.0   

   pregend2  nbrnaliv  multbrth  ...  laborfor_i  religion_i  metro_i  \
0       NaN       1.0       NaN  ...           0           0        0   
1       NaN       1.0       NaN  ...           0           0        0   
2       NaN       3.0       5.0  ...           0           0        0   
3       NaN       1.0       NaN  ...           0           0        0   
4       NaN       1.0       NaN  ...           0           0        0   

       basewgt  adj_mod_basewgt      finalwgt  secu_p  sest  cmintvw  \
0  3410.389399      3869.349602   6448.271112       2     9      NaN   
1  3410.389399      3869.349602   6448.271112       2     9      NaN   
2  7226.301740      8567.549110  12999.542264       2    12      NaN   
3  7226.301740      8567.549110  12999.542264       2    12      NaN   
4  7226.301740      8567.549110  12999.542264       2    12      NaN   

   totalwgt_lb  
0       8.8125  
1       7.8750  
2       9.1250  
3       7.0000  
4       6.1875  

[5 rows x 244 columns]

Print the column names.

In [3]:
preg.columns

Index(['caseid', 'pregordr', 'howpreg_n', 'howpreg_p', 'moscurrp', 'nowprgdk',
       'pregend1', 'pregend2', 'nbrnaliv', 'multbrth',
       ...
       'laborfor_i', 'religion_i', 'metro_i', 'basewgt', 'adj_mod_basewgt',
       'finalwgt', 'secu_p', 'sest', 'cmintvw', 'totalwgt_lb'],
      dtype='object', length=244)

Select a single column name.

In [5]:
preg.columns[10]

'cmotpreg'

Select a column and check what type it is.

In [9]:
pregordr = preg['pregordr']
type(pregordr)

pandas.core.series.Series

Print a column.

In [10]:
pregordr

0        1
1        2
2        1
3        2
4        3
5        1
6        2
7        3
8        1
9        2
10       1
11       1
12       2
13       3
14       1
15       2
16       3
17       1
18       2
19       1
20       2
21       1
22       2
23       1
24       2
25       3
26       1
27       1
28       2
29       3
        ..
13563    2
13564    3
13565    1
13566    1
13567    1
13568    2
13569    1
13570    2
13571    3
13572    4
13573    1
13574    2
13575    1
13576    1
13577    2
13578    1
13579    2
13580    1
13581    2
13582    3
13583    1
13584    2
13585    1
13586    2
13587    3
13588    1
13589    2
13590    3
13591    4
13592    5
Name: pregordr, Length: 13593, dtype: int64

Select a single element from a column.

In [11]:
pregordr[0]

1

Select a slice from a column.

In [12]:
pregordr[2:5]

2    1
3    2
4    3
Name: pregordr, dtype: int64

Select a column using dot notation.

In [13]:
pregordr = preg.pregordr

Count the number of times each value occurs.

In [14]:
preg.outcome.value_counts().sort_index()

1    9148
2    1862
3     120
4    1921
5     190
6     352
Name: outcome, dtype: int64

Check the values of another variable.

In [15]:
preg.birthwgt_lb.value_counts().sort_index()

0.0        8
1.0       40
2.0       53
3.0       98
4.0      229
5.0      697
6.0     2223
7.0     3049
8.0     1889
9.0      623
10.0     132
11.0      26
12.0      10
13.0       3
14.0       3
15.0       1
Name: birthwgt_lb, dtype: int64

Make a dictionary that maps from each respondent's `caseid` to a list of indices into the pregnancy `DataFrame`.  Use it to select the pregnancy outcomes for a single respondent.

In [16]:
caseid = 10229
preg_map = nsfg.MakePregMap(preg)
indices = preg_map[caseid]
preg.outcome[indices].values

array([4, 4, 4, 4, 4, 4, 1], dtype=int64)

## Exercises

Select the `birthord` column, print the value counts, and compare to results published in the [codebook](http://www.icpsr.umich.edu/nsfg6/Controller?displayPage=labelDetails&fileCode=PREG&section=A&subSec=8016&srtLabel=611933)

In [17]:
# Solution goes here

We can also use `isnull` to count the number of nans.

In [18]:
preg.birthord.isnull().sum()

4445

Select the `prglngth` column, print the value counts, and compare to results published in the [codebook](http://www.icpsr.umich.edu/nsfg6/Controller?displayPage=labelDetails&fileCode=PREG&section=A&subSec=8016&srtLabel=611931)

In [19]:
# Solution goes here

To compute the mean of a column, you can invoke the `mean` method on a Series.  For example, here is the mean birthweight in pounds:

In [20]:
preg.totalwgt_lb.mean()

7.265628457623368

Create a new column named <tt>totalwgt_kg</tt> that contains birth weight in kilograms.  Compute its mean.  Remember that when you create a new column, you have to use dictionary syntax, not dot notation.

In [21]:
# Solution goes here

`nsfg.py` also provides `ReadFemResp`, which reads the female respondents file and returns a `DataFrame`:

In [22]:
resp = nsfg.ReadFemResp()

`DataFrame` provides a method `head` that displays the first five rows:

In [23]:
resp.head()

caseid  rscrinf  rdormres  rostscrn  rscreenhisp  rscreenrace  age_a  \
0    2298        1         5         5            1          5.0     27   
1    5012        1         5         1            5          5.0     42   
2   11586        1         5         1            5          5.0     43   
3    6794        5         5         4            1          5.0     15   
4     616        1         5         4            1          5.0     20   

   age_r  cmbirth  agescrn  ...  pubassis_i      basewgt  adj_mod_basewgt  \
0     27      902       27  ...           0  3247.916977      5123.759559   
1     42      718       42  ...           0  2335.279149      2846.799490   
2     43      708       43  ...           0  2335.279149      2846.799490   
3     15     1042       15  ...           0  3783.152221      5071.464231   
4     20      991       20  ...           0  5341.329968      6437.335772   

      finalwgt  secu_r  sest  cmintvw  cmlstyr  screentime   intvlngth  
0  5556.717241       2    18     1234     1222    18:26:36  110.492667  
1  4744.191350       2    18     1233     1221    16:30:59   64.294000  
2  4744.191350       2    18     1234     1222    18:19:09   75.149167  
3  5923.977368       2    18     1234     1222    15:54:43   28.642833  
4  7229.128072       2    18     1233     1221    14:19:44   69.502667  

[5 rows x 3087 columns]

Select the `age_r` column from `resp` and print the value counts.  How old are the youngest and oldest respondents?

In [24]:
# Solution goes here

We can use the `caseid` to match up rows from `resp` and `preg`.  For example, we can select the row from `resp` for `caseid` 2298 like this:

In [25]:
resp[resp.caseid==2298]

caseid  rscrinf  rdormres  rostscrn  rscreenhisp  rscreenrace  age_a  \
0    2298        1         5         5            1          5.0     27   

   age_r  cmbirth  agescrn  ...  pubassis_i      basewgt  adj_mod_basewgt  \
0     27      902       27  ...           0  3247.916977      5123.759559   

      finalwgt  secu_r  sest  cmintvw  cmlstyr  screentime   intvlngth  
0  5556.717241       2    18     1234     1222    18:26:36  110.492667  

[1 rows x 3087 columns]

And we can get the corresponding rows from `preg` like this:

In [26]:
preg[preg.caseid==2298]

caseid  pregordr  howpreg_n  howpreg_p  moscurrp  nowprgdk  pregend1  \
2610    2298         1        NaN        NaN       NaN       NaN       6.0   
2611    2298         2        NaN        NaN       NaN       NaN       6.0   
2612    2298         3        NaN        NaN       NaN       NaN       6.0   
2613    2298         4        NaN        NaN       NaN       NaN       6.0   

      pregend2  nbrnaliv  multbrth  ...  laborfor_i  religion_i  metro_i  \
2610       NaN       1.0       NaN  ...           0           0        0   
2611       NaN       1.0       NaN  ...           0           0        0   
2612       NaN       1.0       NaN  ...           0           0        0   
2613       NaN       1.0       NaN  ...           0           0        0   

          basewgt  adj_mod_basewgt     finalwgt  secu_p  sest  cmintvw  \
2610  3247.916977      5123.759559  5556.717241       2    18      NaN   
2611  3247.916977      5123.759559  5556.717241       2    18      NaN   
2612  3247.916977      5123.759559  5556.717241       2    18      NaN   
2613  3247.916977      5123.759559  5556.717241       2    18      NaN   

      totalwgt_lb  
2610       6.8750  
2611       5.5000  
2612       4.1875  
2613       6.8750  

[4 rows x 244 columns]

How old is the respondent with `caseid` 1?

In [23]:
# Solution goes here

What are the pregnancy lengths for the respondent with `caseid` 2298?

In [24]:
# Solution goes here

What was the birthweight of the first baby born to the respondent with `caseid` 5012?

In [25]:
# Solution goes here